### PR TITLE
Disable shadow host suffix append

### LINKED ISSUE #9094: Mirrored traffic has `-shadow` appended to the hostname in the Host header breaking routing in downstream proxies
*Title*: *Cannot route traffic shadowed by envoy due to addition of `-shadow` to Host header*

*Description*:
The Host and Authority header are modified when shadowing traffic, even when using `host_rewrite` to modify Host manually on the route config.

This results in mirrored requested being dropped by reverse proxies further downstream.

*Example Config*:

Main config ( /etc/envoy/config.yaml ):
```
static_resources:
  listeners:
  - address:
      socket_address:
        address: 0.0.0.0
        port_value: 8080
    filter_chains:
    - filters:
      - name: envoy.http_connection_manager
        config:
          codec_type: auto
          stat_prefix: ingress_http
          rds:
            route_config_name: local_route
            config_source:
              path: /etc/envoy/config.route.yaml
          http_filters:
          - name: envoy.router
            config: {}
  clusters:
  - name: target-service
    connect_timeout: 15s
    type: LOGICAL_DNS
    dns_lookup_family: V4_ONLY
    lb_policy: round_robin
    hosts:
      - socket_address:
          address: localhost
          port_value: 8081
  - name: mirror-service
    connect_timeout: 15s
    type: LOGICAL_DNS
    dns_lookup_family: V4_ONLY
    lb_policy: round_robin
    hosts:
      - socket_address:
          address: foobar.new-cluster.company.com
          port_value: 443
    tls_context: { sni: foobar.new-cluster.company.com }

node:
  cluster: cluster0
  id: node0
```

Route config ( /etc/envoy/config.route.yaml ):
```
version_info: "0"
resources:
- "@type": type.googleapis.com/envoy.api.v2.RouteConfiguration
  name: local_route
  virtual_hosts:
  - name: target-service
    domains:
    - "*"
    routes:
    - match:
        prefix: "/"
      route:
        cluster: target-service
        host_rewrite: foobar.new-cluster.company.com
        request_mirror_policy:
          cluster: mirror-service
          runtime_fraction:
            default_value:
              numerator: 34
```

*What happens*:
1. Request comes into envoy for host (target host "foobar.old-cluster.company.com" with Header: "HOST: foobar.old-cluster.company.com")
2. Request is forwarded directly to ip of the server (e.g.: localhost:8081 with Header: "HOST: foobar.new-cluster.company.com")
3. Request is mirrored to mirror service (e.g.: foobar.new-cluster.company.com:443 with Header: "HOST: foobar.new-cluster.company.com-shadow")
4. Request is dropped by downstream ELB as the Host is invalid.

*Expected behaviour*:
1. Request comes into envoy for host (target host "foobar.old-cluster.company.com" with Header: "HOST: foobar.old-cluster.company.com")
2. Request is forwarded directly to ip of the server (e.g.: localhost:8081 with Header: "HOST: foobar.new-cluster.company.com")
3. Request is mirrored to mirror service (e.g.: foobar.new-cluster.company.com:443 with Header: "HOST: foobar.new-cluster.company.com")
4. Request is routed to mirror service downstream correctly.

*Relevant Links*:
* Code causing issue: https://github.com/envoyproxy/envoy/blob/b9f798dd1dec86d3ed067c449e3c3a49c57984b1/source/common/router/shadow_writer_impl.cc#L30

*Potential Solution*:

We could solve this by adding an overriding host_rewrite in the `request_mirror_policy` config. As an example:

```
...
        request_mirror_policy:
          cluster: mirror-service
          host_rewrite: "foobar.new-cluster.company.com" # (Default: "" would just append shadow)
          runtime_fraction:
            default_value:
              numerator: 34
...
```

### PR DESCRIPTION
**Pull request by jh125486**
Commit Message:
Add runtime feature envoy.reloadable_features.enable_shadow_header_suffix to toggle appending the -shadow suffix to mirrored clusters. Default is false (appends -shadow).

Additional Description: Issue: https://github.com/envoyproxy/envoy/issues/9094

Risk Level: Low

Testing: Unit

Docs Changes: Updated route mirror

Release Notes: Updated current.yaml

Platform Specific Features: n/a

Optional Runtime guard: `envoy_reloadable_features_enable_shadow_header_suffix`

Fixes #9094

### CODE REVIEW

File: source/common/router/shadow_writer_impl.cc [hunk 1]
```diff
@@ -58,7 +59,9 @@ Upstream::ThreadLocalCluster* ShadowWriterImpl::getClusterAndPreprocessHeadersAn
     return nullptr;
   }
 
-  headers.setHost(shadowAppendedHost(headers.getHostValue()));
+  if (Runtime::runtimeFeatureEnabled("envoy.reloadable_features.enable_shadow_header_suffix")) {
```
Comment (botengyao) [inline source/common/router/shadow_writer_impl.cc:62]: I'd think this can be a shadow policy config rather than a runtime guard for this per the comment from the issue.

e.g., we can add `disable_shadow_host_suffix_append` to RequestMirrorPolicy. It is false by default for backward adoption.

https://github.com/envoyproxy/envoy/blob/8879434500ba7e76ff27d41087939bfab635f358/api/envoy/config/route/v3/route_components.proto#L772

**Review by botengyao — COMMENTED** [review]
Thanks for working on this!

File: source/common/router/shadow_writer_impl.cc (hunk 1 above)
Comment (jh125486) [reply source/common/router/shadow_writer_impl.cc:62]: OK, I had worked with a few people in the Slack and this was the direction that was given... I'll do the proto change instead.

**Comment by zuercher** [comment]
/wait

Unless the plan is to remove the suffix entirely in all cases, I think the proto change is more in line with how we usually handle conditional features. Especially since this could useful on a cluster-by-cluster basis.

**Comment by jh125486** [comment]
I’m working on the proto change/tests, a bit slow because building is broken on ARM, but I should have something tomorrow morning.

V/r
Jacob Hochstetler
________________________________

**Comment by jh125486** [comment]
> /wait
> 
> Unless the plan is to remove the suffix entirely in all cases, I think the proto change is more in line with how we usually handle conditional features. Especially since this could useful on a cluster-by-cluster basis.

Ready for review on the proto API changes.

File: source/common/router/config_impl.h [hunk 2]
```diff
@@ -474,13 +474,15 @@ class ShadowPolicyImpl : public ShadowPolicy {
...
   bool traceSampled() const override { return trace_sampled_; }
+  bool disableShadowHostSuffixAppend() const override { return disable_shadow_host_suffix_append_; }
 
 private:
   const std::string cluster_;
   const Http::LowerCaseString cluster_header_;
   std::string runtime_key_;
   envoy::type::v3::FractionalPercent default_value_;
   bool trace_sampled_;
+  bool disable_shadow_host_suffix_append_;
```
Comment (adisuissa) [inline source/common/router/config_impl.h:485]: nit: consider converting to const, and maybe moving the initialization to the c'tor's MIL.

File: api/envoy/config/route/v3/route_components.proto [hunk 3]
```diff
@@ -814,6 +814,9 @@ message RouteAction {
 
     // Determines if the trace span should be sampled. Defaults to true.
     google.protobuf.BoolValue trace_sampled = 4;
+
+    // Disables appending the ``-shadow`` suffix to the shadowed cluster name. Defaults to false.
+    google.protobuf.BoolValue disable_shadow_host_suffix_append = 6;
```
Comment (adisuissa) [inline api/envoy/config/route/v3/route_components.proto:819]: OOC the field defaults to false, so why not to use plain `bool` type? It may make sense to use a wrapped type if it is expected that the default value will change in the future, but I'm not sure if this is the case.

File: test/common/router/shadow_writer_impl_test.cc [hunk 4]
```diff
@@ -53,18 +56,26 @@ class ShadowWriterImplTest : public testing::Test {
...
 
 TEST_F(ShadowWriterImplTest, Failure) {
   InSequence s;
 
-  expectShadowWriter("cluster1:8000", "cluster1-shadow:8000");
+  expectShadowWriter("cluster1:8000", "cluster1-shadow:8000", false);
   callback_->onFailure(request_, Http::AsyncClient::FailureReason::Reset);
 }
 
+TEST_F(ShadowWriterImplTest, SuccessNoAppend) {
```
Comment (adisuissa) [inline test/common/router/shadow_writer_impl_test.cc:71]: nit: change the name of the test to something more understandable, e.g., SuccessAvoidShadowHeaderPrefixChange.
Also feel free to add a one-line comment that describes the test.

File: api/envoy/config/route/v3/route_components.proto [hunk 5]
```diff
@@ -814,6 +814,9 @@ message RouteAction {
 
     // Determines if the trace span should be sampled. Defaults to true.
     google.protobuf.BoolValue trace_sampled = 4;
+
+    // Disables appending the ``-shadow`` suffix to the shadowed cluster name. Defaults to false.
```
Comment (adisuissa) [inline api/envoy/config/route/v3/route_components.proto:818]: Should this be:
```suggestion
    // Disables appending the ``-shadow`` suffix to the shadowed host header. Defaults to false.
```

File: source/common/router/shadow_writer_impl.cc [hunk 6]
```diff
@@ -58,7 +58,12 @@ Upstream::ThreadLocalCluster* ShadowWriterImpl::getClusterAndPreprocessHeadersAn
     return nullptr;
   }
 
-  headers.setHost(shadowAppendedHost(headers.getHostValue()));
+  // Append "-shadow" suffix if option is disabled.
+  if (const_cast<Http::AsyncClient::RequestOptions&>(options).is_shadow_suffixed_disabled) {
```
Comment (adisuissa) [inline source/common/router/shadow_writer_impl.cc:62]: Why is the const-case needed here?

File: source/common/router/shadow_writer_impl.cc [hunk 7]
```diff
@@ -58,7 +58,12 @@ Upstream::ThreadLocalCluster* ShadowWriterImpl::getClusterAndPreprocessHeadersAn
...
+    headers.setHost(headers.getHostValue());
```
Comment (adisuissa) [inline source/common/router/shadow_writer_impl.cc:63]: Why is this line needed?
IIUC it just does: `headers.host = headers.host;`, no?

**Review by adisuissa — COMMENTED** [review]
Thank you for your contribution! Overall LGTM.
In addition to the comments, please add an integration test that tests this feature and a release-note.

/wait

**Comment by jh125486** [comment]
> Thank you for your contribution! Overall LGTM. In addition to the comments, please add an integration test that tests this feature and a release-note.
> 
> /wait

Yep, I can fix these things.  
For release notes, I thought the section I added to `current.yaml` was where it went. Is there another place I need to put release notes?

For full transparency, I can't get the dev container to work through our proxy, bazel arm builds are broken  (`The current user has CAP_DAC_OVERRIDE set` error) and consequently I'm sorta flying blind without an IDE.

File: api/envoy/config/route/v3/route_components.proto (hunk 3 above)
Comment (jh125486) [reply api/envoy/config/route/v3/route_components.proto:819]: I used the well-known type to avoid a `false` getting marshaled for that field.

Comment (htuch) [reply api/envoy/config/route/v3/route_components.proto:819]: What does getting marshaled refer to here?

Comment (jh125486) [reply api/envoy/config/route/v3/route_components.proto:819]: Whenever this proto is marshaled to anything... JSON, XML, etc.
If it's a raw `bool`, then the field will be explicitly marshaled... otherwise if it's not set, the `null` would be omitted?

Essentially I just used the WKT to avoid any side-effects.

File: source/common/router/config_impl.h (hunk 2 above)
Comment (jh125486) [reply source/common/router/config_impl.h:485]: So want me to add it to `RequestMirrorPolicy` in the MIL or as another parameter to the MIL?

**Comment by phlax** [comment]
> I thought the section I added to current.yaml was where it went

all good - it is


> For full transparency, I can't get the dev container to work through our proxy, bazel arm builds are broken (The current user has CAP_DAC_OVERRIDE set error) and consequently I'm sorta flying blind without an IDE.

looking at the `.devcontainer` caps - i think its overly permissive - not sure why it needs so much capability - iirc these can be used for privilege escalation

https://github.com/envoyproxy/envoy/blob/45ab9cfa30c7631fe325847dd291e2dcc80c1894/.devcontainer/devcontainer.json#L6-L9

i dont use vscode myself, so not sure if they are necessary - my guess is that these were just copied from the `ci/run_envoy_docker.sh` script which i think used to have similar

**Comment by phlax** [comment]
windows failiure is unrelated https://github.com/envoyproxy/envoy/issues/32060

File: api/envoy/config/route/v3/route_components.proto (hunk 5 above)
Comment (phlax) [inline api/envoy/config/route/v3/route_components.proto:818]:
```suggestion
    // Disables appending the ``-shadow`` suffix to the shadowed ``Host`` header. Defaults to ``false``.
```

formatting nits

**Comment by jh125486** [comment]
Sorry it took so long... got the requests changes done.  It's currently working through my side.

**Comment by jh125486** [comment]
> This pull request has been automatically marked as stale because it has not had activity in the last 30 days. It will be closed in 7 days if no further activity occurs. Please feel free to give a status update now, ping for review, or re-open when it's ready. Thank you for your contributions!

Still currently working through my side.

**Comment by jh125486** [comment]
@phlax PR has been updated, sorry for the delay.
I have *not* updated the change log yet, as it didn't make sense to do so until the actual code was looked over.

**Comment by RyanTheOptimist** [comment]
/assign @htuch 
Harvey it looks like you've already taken a look at this PR. Mind reviewing it?

File: api/envoy/config/route/v3/route_components.proto (hunk 3 above)
Comment (htuch) [reply api/envoy/config/route/v3/route_components.proto:819]: I don't think that's how proto3 -> JSON translation works by default, see https://protobuf.dev/programming-guides/proto3/#json:

> Always emit fields without presence: Fields that don’t support presence and that have their default value are omitted by default in JSON output (for example, an implicit presence integer with a 0 value, implicit presence string fields that are empty strings, and empty repeated and map fields). An implementation may provide an option to override this behavior and output fields with their default values.

Envoy API style prefers to use scalar for these cases.

Comment (jh125486) [reply api/envoy/config/route/v3/route_components.proto:819]: Huh.  
I was under the impression that WKTs were used so that they were omitted from the JSON when not set, and scalar fields are always serialized.

Comment (htuch) [reply api/envoy/config/route/v3/route_components.proto:819]: No, WKTs are used to support the testing of absence / presence to provide the ability to provide default values for example.

**Review by adisuissa — COMMENTED** [review]
Thanks!
/lgtm api

**Review by htuch — CHANGES_REQUESTED** [review]
LGTM, can you add a release note at https://www.envoyproxy.io/docs/envoy/latest/version_history/v1.31/v1.31.0? Thanks.

**Review by adisuissa — APPROVED** [review]
LGTM, thanks!
/lgtm api

**Review by htuch — APPROVED** [review]
